discovery/consul.go

```diff
@@ -0,0 +1,61 @@
+package discovery
+
+import (
+	"bytes"
+	"errors"
+	"net/url"
+	"strconv"
+
+	"github.com/hashicorp/consul/api"
+)
+
+// Consul is a Discoverer that uses Consul to find
+// healthy instances of a given name.
+type Consul struct {
+	ConsulHealth *api.Health
+}
+
+// NewConsul creates a new instance of a Consul Discoverer
+func NewConsul(config *api.Config) (*Consul, error) {
+	consulClient, err := api.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Consul{
+		ConsulHealth: consulClient.Health(),
+	}, nil
+}
+
+// UpdateDestinations updates the list of destinations based on healthy nodes
+// found via Consul.
+func (c *Consul) UpdateDestinations(serviceName string) ([]string, error) {
+	serviceEntries, _, err := c.ConsulHealth.Service(serviceName, "", true, &api.QueryOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	numHosts := len(serviceEntries)
+	if numHosts < 1 {
+		return nil, errors.New("Received no hosts from Consul")
+	}
+	// Make a slice to hold our returned hosts
+	hosts := make([]string, numHosts)
+	for index, se := range serviceEntries {
+		service := se.Service
+
+		var h bytes.Buffer
+		h.WriteString(se.Node.Address)
+		h.WriteString(":")
+		h.WriteString(strconv.Itoa(service.Port))
```

Comment:
yeah, building up the bytes buffer is slightly faster! on my machine:
```
BenchmarkSprintf-8   	10000000	       232 ns/op
BenchmarkBuffer-8    	10000000	       121 ns/op
```
running
```
func BenchmarkSprintf(b *testing.B) {
	for i := 0; i < b.N; i++ {
		fmt.Sprintf("%s:%d", "nodeaddress", 8000)
	}
}
func BenchmarkBuffer(b *testing.B) {
	for i := 0; i < b.N; i++ {
		var h bytes.Buffer
		h.WriteString("nodeaddress")
		h.WriteString(":")
		h.WriteString(strconv.Itoa(8000))
	}
}
```